src/sync/app.cpp

```diff
@@ -0,0 +1,163 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or utilied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "app.hpp"
+#include <sstream>
+#include "sync_manager.hpp"
+#include "generic_network_transport.hpp"
+#include "app_credentials.hpp"
+#include "../external/json/json.hpp"
+
+// wrap an optional json key into the Optional type
+#define WRAP_JSON_OPT(JSON, KEY, RET_TYPE) \
+JSON.find(KEY) != JSON.end() ? Optional<RET_TYPE>(JSON[KEY].get<RET_TYPE>()) : realm::util::none
+
+namespace realm {
+namespace app {
+
+std::shared_ptr<App> App::app(const std::string app_id,
+                              const realm::util::Optional<App::Config> config)
+{
+    return std::make_shared<App>(app_id, config);
+}
+
+
+
+static std::unique_ptr<error::AppError> handle_error(const Response response) {
```

Comment:
With the introduction of `custom_error_code` needs to be rewritten I think? As a minimum, we should send the custom error back + message back to bindings. 

Also returning `{}` seems like it would be very hard for bindings to figure out exactly what is going wrong. IMO we should probably alway send on the error code and the message (which can be parsed if relevant).